### ISSUE TITLE
[GC performance] The performance of v2 manifest deletion is not good in S3 environment

### ISSUE THREAD

**Issue by wy65701436** [goharbor/harbor#12948]
In S3 backend environment, we found that it took about 39 seconds to delete a manifest via v2 API.

**[why still use v2 to handle manifest deletion]**

As Harbor cannot know the tags belong to the manifest in the storage, so the GC job needs to leverage the v2 API to remove them. But, the v2 API will look up all of tags, remove them one by one. This may cause performance issue.

[what we can do next]

1, Investigate how many requests send to S3 storage within the v2 manifest deletion.
2, Investigate the possibility of not to store the _first tag_ in the backend, then GC job can skip this step.

**Log**
```
Sep 1 12:56:35 192.168.144.1 registry[1146]: time="2020-09-01T12:56:35.750530108Z" level=info msg="authorized request" go.version=go1.13.8 http.request.host="registry:5000" http.request.id=c9a4d5ad-4157-4091-a023-93d8e20a5746 http.request.method=DELETE http.request.remoteaddr="192.168.144.9:44072" http.request.uri="/v2/library/testingg/manifests/sha256:20f39c20df7c5605f77862b711c3d28731e4d569171ec852ce34a06432611faa" http.request.useragent=harbor-registry-client vars.name="library/testingg" 
vars.reference="sha256:20f39c20df7c5605f77862b711c3d28731e4d569171ec852ce34a06432611faa" 
Sep 1 12:57:14 192.168.144.1 registry[1146]: time="2020-09-01T12:57:14.340710966Z" level=info msg="response completed" go.version=go1.13.8 http.request.host="registry:5000" http.request.id=c9a4d5ad-4157-4091-a023-93d8e20a5746 http.request.method=DELETE http.request.remoteaddr="192.168.144.9:44072" http.request.uri="/v2/library/testingg/manifests/sha256:20f39c20df7c5605f77862b711c3d28731e4d569171ec852ce34a06432611faa" http.request.useragent=harbor-registry-client http.response.duration=38.598453034s http.response.status=202 http.response.written=0 
```

**Comment by dkulchinsky** [comment]
Hey @wy65701436, following our chat in Slack I'd like to share similar performance issue we're experiencing with a GCS storage backend.

We're running Harbor v2.1.1 and we replicated a GCR registry content to Harbor, however we forgot to exclude a repo that had at the time >60,000 tags.

After replication completed we deleted the repo in Harbor and ran GC, but the job keeps failing due to timeout to delete manifest:

```
2020-11-03T16:44:21Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:259]: delete the manifest with registry v2 API: fastly/demo-go-app, application/vnd.docker.distribution.manifest.v2+json, sha256:0f20ddd7f417cae5d685a3ea653617241135644575744db58cf5091dcd6cdf5c

2020-11-03T17:14:21Z [ERROR] [/jobservice/job/impl/gc/garbage_collection.go:262]: failed to delete manifest with v2 API, fastly/demo-go-app, sha256:0f20ddd7f417cae5d685a3ea653617241135644575744db58cf5091dcd6cdf5c, Delete "http://harbor-registry:5000/v2/fastly/demo-go-app/manifests/sha256:0f20ddd7f417cae5d685a3ea653617241135644575744db58cf5091dcd6cdf5c": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

2020-11-03T17:14:21Z [ERROR] [/jobservice/job/impl/gc/garbage_collection.go:165]: failed to execute GC job at sweep phase, error: failed to delete manifest with v2 API: fastly/demo-go-app, sha256:0f20ddd7f417cae5d685a3ea653617241135644575744db58cf5091dcd6cdf5c: Delete "http://harbor-registry:5000/v2/fastly/demo-go-app/manifests/sha256:0f20ddd7f417cae5d685a3ea653617241135644575744db58cf5091dcd6cdf5c": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
```

looking at the registry logs we see that it takes over an hour to delete a manifest:
```
[harbor-registry-b4fbbb8df-xcgt4 registry] 127.0.0.1 - - [03/Nov/2020:16:44:21 +0000] "DELETE /v2/fastly/demo-go-app/manifests/sha256:0f20ddd7f417cae5d685a3ea653617241135644575744db58cf5091dcd6cdf5c HTTP/1.1" 202 0 "" "harbor-registry-client"

[harbor-registry-b4fbbb8df-xcgt4 registry] time="2020-11-03T17:56:14.798548519Z" level=info msg="response completed" go.version=go1.14.7 http.request.host="harbor-registry:5000" http.request.id=3b01c244-6b14-49ba-bfde-1bfd7934c15e http.request.method=DELETE http.request.remoteaddr="127.0.0.1:59854" http.request.uri="/v2/fastly/demo-go-app/manifests/sha256:0f20ddd7f417cae5d685a3ea653617241135644575744db58cf5091dcd6cdf5c" http.request.useragent=harbor-registry-client http.response.duration=1h11m53.181188629s http.response.status=202 http.response.written=0
```

We enabled debug log in registry and we saw it was spending most of the time iterating though the tags with `gcs.GetContent`:
```
❯ grep GetContent harbor-registry-debug-logs.txt|grep demo-go-app|wc -l
   55525
```

for example:
```
[harbor-registry-b4fbbb8df-xcgt4 registry] time="2020-11-03T16:44:27.205475289Z" level=debug msg="gcs.GetContent("/docker/registry/v2/repositories/fastly/demo-go-app/_manifests/tags/0023fd10bee5f3cc968e55148169091eb7d1cf795a8780ee7508642ab047042b/current/link")" auth.user.name="harbor_registry_user" go.version=go1.14.7 http.request.host="harbor-registry:5000" http.request.id=3b01c244-6b14-49ba-bfde-1bfd7934c15e http.request.method=DELETE http.request.remoteaddr="127.0.0.1:59854" http.request.uri="/v2/fastly/demo-go-app/manifests/sha256:0f20ddd7f417cae5d685a3ea653617241135644575744db58cf5091dcd6cdf5c" http.request.useragent=harbor-registry-client trace.duration=110.978462ms trace.file="/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).GetContent" trace.id=6374bc55-1500-4fb9-bcf0-f53da9f5fa16 trace.line=95 vars.name="fastly/demo-go-app" vars.reference="sha256:0f20ddd7f417cae5d685a3ea653617241135644575744db58cf5091dcd6cdf5c"
```

I can share the full GC job and registry debug logs if needed, also happy to provide more information.

**Comment by guyguy333** [comment]
Experiencing exactly the same issue as @dkulchinsky. We're unable to end a GC. It always end with a context deadline exceeded. We've more than 1Tb to clean (~130k objects). We can't resume so we have to start from scratch again.

**Comment by dkulchinsky** [comment]
Hello friends, I'd like to ask to raise the priority of this issue.

We are running several instances of Harbor (we use GCS backend, but I think the root cause here is the same) and we're rapidly growing our usage.

We are starting to reach capacities that the GC simply cannot handle, repositories with more than a few thousand tags is taking ~2 minutes to delete a single manifest during GC, GC is now taking 10~14 hours and the problem is getting worse every day since we're adding more tags then we are deleting.

on our test/certification Harbor instance we've reached over 20,000 tags on some repositories and GC just times out on the first manifest since the lookup takes >20 minutes.

We are concerned about increasing our storage costs since we can't clean it up as well as other potential issues that may arise from having all these blobs/manifests lingering with no ability to properly clean them up.

this issue was tagged as a candidate for v2.2.0, and we're already seeing v2.4.0 going out the door.


I'm happy to provide additional context, information, logs but just hope we can some attention on this issues since I think it will impact any user that needs Harbor to work at scale.

/cc @wy65701436 @reasonerjt

**Comment by dkulchinsky** [comment]
I believe this is still an active issue being tracked, so probably shouldn't get closed yet.

**Comment by rcjames** [comment]
This is still an issue.

**Comment by dkulchinsky** [comment]
@wy65701436 any hints on when the team may find some time to look at this? this seems like an issue that requires desperate attention however it didn't see any traction in over 2 years now.

**Comment by twhiteman** [comment]
Just to further explain the crux of this issue.

Harbor is using the docker distribution for it's (harbor-registry) registry component.

The harbor GC will call into the docker registry to delete the manifest, which in turn then does a lookup for all tags that reference the deleted manifest:
https://github.com/distribution/distribution/blob/78b9c98c5c31c30d74f9acb7d96f98552f2cf78f/registry/handlers/manifests.go#L536

To find the tag references, the docker registry will iterate *every* tag in the repository and read it's link file to check if it matches the deleted manifest (i.e. to see if uses the same sha256 digest):
https://github.com/distribution/distribution/blob/78b9c98c5c31c30d74f9acb7d96f98552f2cf78f/registry/storage/tagstore.go#L160

So, the more tags you have in your repository, the worse the performance will be (as there will be more s3 API calls occurring for the tag directory lookups and tag file reads).

**Comment by captn3m0** [comment]
not stale.

**Comment by Antiarchitect** [comment]
Seems facing this in https://github.com/goharbor/harbor/issues/17536#issuecomment-1399356186 Extremely slow.

**Comment by karmicdude** [comment]
Should we expect a workaround on this issue? We have about ~2.6TB of storage on minio tenant. Running retention and GC just doesn't work - in fact with the deletion speed we have now, it will take years (from about 2 years+). We are considering migrating just PVC based on ssd disk type, but this is hardly any the true way, because it does't give us enough reliability.

**Comment by Antiarchitect** [comment]
@karmicdude This could be a possible fix for that https://github.com/distribution/distribution/pull/3827
Harbor uses distribution under the hood and I already tested it for my Harbor instance

**Comment by dkulchinsky** [comment]
> @karmicdude This could be a fix for that [distribution/distribution#3827](https://github.com/distribution/distribution/pull/3827) Harbor uses distribution under the hood and I already tested it for my Harbor instance

wow, that's very encouraging @Antiarchitect! hope your PR will be accepted, happy to test to myself if you perhaps have a pre-built image posted somewhere I can use? or can try build one myself.

**Comment by Antiarchitect** [comment]
@dkulchinsky Image is in our private registry, but the instruction is this:
Build distrubution bin from my fork via `go build -o bin/registry ./cmd/registry`
And place the bin inside `harbor-registry` folder of this slightly modified bitnami repo 
[harbor-registry.tar.gz](https://github.com/goharbor/harbor/files/10794061/harbor-registry.tar.gz)

Then docker build

**Comment by dkulchinsky** [comment]
> @dkulchinsky Image is in our private registry, but the instruction is this: Build distrubution bin from my fork via `go build -o bin/registry ./cmd/registry` And place the bin inside `harbor-registry` folder of this slightly modified bitnami repo [harbor-registry.tar.gz](https://github.com/goharbor/harbor/files/10794061/harbor-registry.tar.gz)
> 
> Then docker build

Thanks for this @Antiarchitect! looks like the only issue with above is the manual redis sentinel patch that Harbor folks are applying during the build process: https://github.com/goharbor/harbor/blob/4141d774070e4fa050ed21a063be0a50745beb5a/make/photon/registry/builder#L31-L34

this patch no longer works against distribution's main branch as-is, trying to see if I can get this sorted.

**Comment by Antiarchitect** [comment]
@dkulchinsky Please tell about the result. In my case speedup was drastical.

**Comment by dkulchinsky** [comment]
> @dkulchinsky Please tell about the result. In my case speedup was drastical.

Hey @Antiarchitect, here's some stats from last night.

Job started at `2023-02-22T01:45:35Z` and stopped at `2023-02-22T14:14:56Z` so just over 12 hours (we store logs in DB, so I had to stop it to be able to see the logs)

it identified `11516 blobs and 6059 manifests eligible for deletion`, which is weird too, because the dry run I ran just before this one showed `92595 blobs and 51015 manifests eligible for deletion` 🤷🏼‍♂️ 

in 12 hours it processed `[2480/17575]` artifacts

this is definitely an improvement, but still feels like this is quite slow (also, I don't understand why dry-run vs non-dry-run shows totally different blobs/manifests eligible for deletion)

In my particular case, I see a lot of these errors:
```
2023-02-22T14:14:36Z [INFO] [/jobservice/job/impl/gc/util.go:67]: failed to exec v2DeleteManifest, error: http status code: 500, body: {"errors":[{"code":"UNKNOWN","message":"unknown error","detail":{}}]}
, will retry again after: 100ms

2023-02-22T14:14:37Z [INFO] [/jobservice/job/impl/gc/util.go:67]: failed to exec v2DeleteManifest, error: http status code: 404, body: {"errors":[{"code":"MANIFEST_UNKNOWN","message":"manifest unknown"}]}
, will retry again after: 115.21198ms
```

It seems reasonable to retry a 500, but I think it makes no sense to retry a 404 as that is a permanent condition, so this is likely driving up the time spent on each item, for example:
```
2023-02-22T14:13:53Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:283]: [2480/17575] delete the manifest with registry v2 API: <project>/<repo>/demo-go-app, application/vnd.docker.distribution.manifest.v2+json, sha256:71593b6e361ee36900ce5130152781c4e4911fcb664b79811294ffd06d5240ae

<a bunch of 404 errors>

2023-02-22T14:14:54Z [ERROR] [/jobservice/job/impl/gc/garbage_collection.go:286]: [2480/17575] failed to delete manifest with v2 API, <project>/<repo>/demo-go-app, sha256:71593b6e361ee36900ce5130152781c4e4911fcb664b79811294ffd06d5240ae, retry timeout: http status code: 404, body: {"errors":[{"code":"MANIFEST_UNKNOWN","message":"manifest unknown"}]}
```

so a minute spent trying to delete this manifest that returns a 404, eventually it fails and this manifest is not removed from `artifact_thrash` table, so it will just keep retrying it forever:

```
coredb=# select digest from artifact_trash where digest='sha256:71593b6e361ee36900ce5130152781c4e4911fcb664b79811294ffd06d5240ae';
                                 digest
-------------------------------------------------------------------------
 sha256:71593b6e361ee36900ce5130152781c4e4911fcb664b79811294ffd06d5240ae
(1 row)
```

so, bottom line I think this improvement makes total sense and definitely improves things, but there's additional work required to improve GC perfromance overall.

It doesn't seem like current handling of the 404 error makes sense as this would result in never ending cycle of trying to delete these artifacts.

Additionally, we use GCS as our storage backend and the Google SDK in distribution wasn't updated since 2015, so I won't be surprised that many of the above issues we're seeing are simply because the library being used is soooo old and possibly broken at this point. I saw a [PR](https://github.com/distribution/distribution/pull/3702) in flight from last year to update it, but looks like it's stalled.

/cc @wy65701436

**Comment by Antiarchitect** [comment]
@dkulchinsky 
> in 12 hours it processed [2480/17575] artifacts
That was the result before any speedup (approximately). But I have no GCS - could be the case. We have Ceph RadosGW as an S3 compatible backend for blobs storage

**Comment by dkulchinsky** [comment]
> @dkulchinsky
> 
> > in 12 hours it processed [2480/17575] artifacts
> 
> That was the result before any speedup (approximately). But I have no GCS - could be the case. We have Ceph RadosGW as an S3 compatible backend for blobs storage

@Antiarchitect quite likely it's a combination of issues indeed, I could have messed up backporting your PR to v2.8.1, although it seems to be working.

otoh, I think I'm seeing an issue with memory usage since upgrading to this version in sandbox:
![image](https://user-images.githubusercontent.com/61992380/221358890-51a9cd48-696f-4f0d-b9e4-d73f9144a361.png)

this steady increase in memory is happening even when GC job is not running, eventually it OOMs and restarts but you can see it repeats.

**Comment by Antiarchitect** [comment]
@dkulchinsky Could you set `STORAGE_TAGSTORE_LOOKUP_CONCURRENCY` to some higher value e.g. `256` and see if the perfomance changes. I will try to test the latest version of this patch by myself. Unfortunately it needs time to have significant amount of blobs, eligible for deletion to measure on - I can turn off the GC for several days.

**Comment by Antiarchitect** [comment]
```
2023-02-27T16:25:06Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:250]: 3054 blobs and 790 manifests eligible for deletion
2023-02-27T16:25:06Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:251]: The GC could free up 48159 MB space, the size is a rough estimation.

...

2023-02-27T16:44:19Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:403]: 3054 blobs and 790 manifests are actually deleted
2023-02-27T16:44:19Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:404]: The GC job actual frees up 48159 MB space.
```

~ 19min

But still my experience is irrelevant because I've already cleaned up huge bunch of old tags. Still that is much faster that it was before (I had ~2500 blobs per 24 hours progress before the speedup)

**Comment by dkulchinsky** [comment]
Looks very promising @Antiarchitect, I think in my case the issue is compounded by:
1. the slow and repeat handling of 404 for manifest delete
2. GCS library from 2015

**Comment by karmicdude** [comment]
We have a harbor in K8S and `minio` tenant as s3 storage backend with pool of 4 minio-servers. Drives under each server is hdd 320/120 IOPS read/write and 150MB/s (network drives). 2.6TB total for artifacts in all projects. 
We have the most fat project that had almost 1TB of artifacts and we haven't cleaned it in about a year because of GC issue. In the end, we ran retention only on this project.

The first problem is that it takes about 5 hours to prepare for removal (making a list of blobs eligible for deletion) *pain*

I backported the @Antiarchitect fix to the distribution/distribution registry v2.8.0 used in the latest harbor and manually apply patches for redis / sentinel to make it work with the fix. Made minor edits in [this code](https://github.com/goharbor/harbor/tree/main/make/photon/registry/ ) and build image. I also fixed the harbor-helm chart, to support custom env for the registry container.

At first, I started on the `lookupConcurrencyFactor` of `12`. I didn't really detect any changes at all in speeding up. One step of entire deletion [1/...] takes about ~15-30 minutes. Next time I set `lookupConcurrencyFactor` of `64`. It's a little faster, but it's not what I expected. I tried setting it to `128` at once and it was about twice as fast. 

I tried setting `512` and it caused minio to crash, despite the significantly increased resource limits. Any values greater than 256 cause the removal process to become as slow as without the fix, and after some time the minio goes down (servers begin to fall by timeout), although the I/O load is not critical, within the limits of the used drives. In this mode, minio consumes for each of the servers about 3 times the CPU and `N` amount of memory. `N` because no matter how much we increase the memory limits, minio eats up all. Before the GC, each minio server consumed about 2-3Gi of RAM. I tried allocating 16Gi (in an attempt to see if there will be a real acceleration of deletion) and it takes up everything. I think this is because of page cache. 

This is the situation we have now.

```
INFO [/jobservice/job/impl/gc/garbage_collection.go:250]: 308375 blobs and 257272 manifests eligible for deletion
INFO [/jobservice/job/impl/gc/garbage_collection.go:251]: The GC could free up 1473398 MB space, the size is a rough estimation.
....
```
Now I'm going to start GC again with the `lookupConcurrencyFactor` of `200`. But it still takes about forever :(

One of the main problems is that when GC is running push images begins to act very unstable because of the greatly increasing latency. Users can't push large 600Mb+ images  for hours. It's just that when you push layers, the process hangs and doesn't fall off by timeout.
So far I haven't been able to figure out why this happens, now I'm trying to tune different components, primarily minio. Perhaps it would help to transfer minio to non-networked disks, there is much lower latency and much higher IOPS. 
We are now testing our theory by trying to deploy a copy of minio tenant on another type of fast disks from snapshots. And experiment on it until we find a way to run GC without service degradation for clients.

**Comment by karmicdude** [comment]
Yes, we use the minio-operator and minio `tenant` was created under harbor registry. Now the drives are networked, with the specifications listed above. You can see that there is a problem, when GC starts, i/o wait ~450ms - a lot. 
![image](https://user-images.githubusercontent.com/759980/225253461-b9801d57-42f3-47e8-a9b8-73b0d5bf7801.png)
We will test a similar setup on fast disks (ssd or nvme) with the same data made from snapshots to see how the situation changes when GC runs. I will post when I have some results. But it is unlikely to improve the issue much.
If the situation gets much better, we always have the option to deploy minio on dedicated server with high-performance local disks.

**Comment by dkulchinsky** [comment]
@Antiarchitect, wanted to share some details

> Looks very promising @Antiarchitect, I think in my case the issue is compounded by:
> 
> 1. the slow and repeat handling of 404 for manifest delete
> 2. GCS library from 2015

I was able to build distribution with your PR, https://github.com/distribution/distribution/pull/3702 and the Redis sentinel patch.

I also figured out the issue around the retry loop when a 404 is returned during manifest delete and have a fix for Harbor jobservice here: https://github.com/goharbor/harbor/pull/18386

I'm running a test now with the above in our sandbox Harbor and although it's not breaking any speed records, it is looking much better, will update once I have more concrete numbers.

**Comment by hemanth132** [comment]
When using s3 for storage, during manifest delete, s3 look-up using s3 list and get calls is performed to check all the tags referencing this manifest. For a large repository having lakhs of images, all these tags are read to delete 1 single image.

Instead of relying on distribution to delete the tag folder, since Harbor already maintains the tags referencing a manifest, we can introduce a new api in distribution to delete the tag directory altogether when deleting the artifact and skip tag lookup during the garbage collection step.

I was able to test this out and brought down the registry size from 360TB to less than 30TB by configuring 30 days of retention period. Checkout these prs:
https://github.com/hemanth132/harbor/pull/1
https://github.com/hemanth132/harbor/pull/2
https://github.com/hemanth132/distribution/pull/2

Have done a couple of optimizations like:
1. introducing configurable concurrency in GC
2. Add concurrency limit for retention to avoid db CPU spikes.
3. Comment out deletion of orphan blobs not associated with any artifact in garbage collection step since that rarely happens.

I made the changes against the version 2.7.0 code of Harbor. Now that this is working, will try to raise pr in the main harbor repo in the coming weeks.

Please let me know if this helps for you.

**Comment by dkulchinsky** [comment]
Hey @hemanth132, just a quick shout out that this looks very promising and thanks for your effort.

Would love to see these changes make it into Harbor so we can finally run GC 😅 

/cc @Vad1mo, this is related to our conversation in Slack earlier.

**Comment by Vad1mo** [comment]
@wy65701436, take a look https://github.com/goharbor/harbor/issues/12948#issuecomment-1545296951

**Comment by sebglon** [comment]
Any update?

**Comment by jwojnarowicz** [comment]
Has anyone tested @hemanth132 solution? Or are there any updates from Harbor team regarding the API or GC? @Vad1mo @chlins Bump because it's still an important issue regarding usage with every S3 backend.

**Comment by karmicdude** [comment]
Any update? Still causing a huge pain, GC works out slower than data is added, resulting in having to constantly extend disks

**Comment by microyahoo** [comment]
hi @karmicdude, I have taken over @Antiarchitect's efforts with concurrent lookup and untag in PR https://github.com/distribution/distribution/pull/4329. You can try it and check whether it has improvement, thanks.

**Comment by microyahoo** [comment]
> hi @karmicdude, I have taken over @Antiarchitect's efforts with concurrent lookup and untag in PR [distribution/distribution#4329](https://github.com/distribution/distribution/pull/4329). You can try it and check whether it has improvement, thanks.

hi @karmicdude, @sebglon, @jwojnarowicz @sidewinder12s , https://github.com/distribution/distribution/pull/4329 has already merged, would you please help to try whether it meets expectations.

**Comment by karmicdude** [comment]
Nice, I'll definitely check it out

**Comment by kofj** [comment]
Running GC on s3 like storage is so expensive, our many customers has storage more the 2 PiB and 900 million objects, they want quickly and high performance GC. I am interested in this work.

**Comment by Antiarchitect** [comment]
@kofj FYI It seems like concurrent tag lookup feature kindly picked up by @microyahoo was merged into distribution: https://github.com/distribution/distribution/pull/4329

**Comment by wy65701436** [comment]
> Whilst concurrent tag lookup is slightly better, it does not solve the underlying performance issue (it still has to lookup and read all tag files on the S3 filesystem for the referenced repository).
> 
> For better solutions, I'd like to see one of:
> 
>     1. docker distribution uses a reverse tag lookup system (returns all tags referencing the given sha256 in a particular repository)
> 
>     2. Harbor performs the deletion of the s3 manifest file itself (and no longer calls the docker delete manifest API)

I prefer the option 3, which is harbor doesn't perform the tag deletion at all. Why harbor still leverages this api is because that the tag on pushing still be landed into the distribution side. However, actually, it is not necessary since harbor uses the database to do the CRUD of tag. 

In summary, we can store all the artifacts in tag-less in the distribution side. So then, the deletion is not needed, but we should consider about the existing artifact when we try to update the logic. 

Another quick solution is that, we can give an option for the end user to let them decide whether removing the tag from the backend. It's no harm but generates some garbage in the storage side.

**Comment by snowmanstark** [comment]
@wy65701436 is this fix coming in 2.14.0?

**Comment by toddw-as** [comment]
Option 3 is very interesting (even though the developer in me cringes that it leaves lots of small tag files behind in the S3 storage). I do like the idea of putting this as an configurable option, to at least give users a choice when they hit this.

**Comment by twhiteman** [comment]
Option 3 is very interesting (even though the developer in me cringes that it leaves lots of small tag files behind in the S3 storage). I do like the idea of putting this as an configurable option, to at least give users a choice when they hit this.

**Comment by MalteMagnussen** [comment]
The GC jobs claims it can clean up 2.7 Terabyte of space. Of the total Quota used: 5.56TiB that is 47.84% we can remove with GC. 

```
2025-06-22T00:04:19Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:273]: 149210 blobs and 53384 manifests eligible for deletion
2025-06-22T00:04:19Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:274]: The GC could free up 2664437 MB space, the size is a rough estimation.
```

But it runs for over 24 hours, and does barely anything. 

![Image](https://github.com/user-attachments/assets/94106196-b2c4-41a0-9260-c05188e8271c)

**Comment by microyahoo** [comment]
> The GC jobs claims it can clean up 2.7 Terabyte of space. Of the total Quota used: 5.56TiB that is 47.84% we can remove with GC.
> 
> ```
> 2025-06-22T00:04:19Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:273]: 149210 blobs and 53384 manifests eligible for deletion
> 2025-06-22T00:04:19Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:274]: The GC could free up 2664437 MB space, the size is a rough estimation.
> ```
> 
> But it runs for over 24 hours, and does barely anything.
> 
> ![Image](https://github.com/user-attachments/assets/94106196-b2c4-41a0-9260-c05188e8271c)

hi @MalteMagnussen, It looks like the GC task failed. You can check the logs to confirm the reason for the failure first.

**Comment by MalteMagnussen** [comment]
@microyahoo - We can't get the logs most of the time. 

Most of the time the button just responds with this:

```
{"errors":[{"code":"NOT_FOUND","message":"{\"code\":10010,\"message\":\"object is not found\",\"details\":\"7584cbdfcd1932b7f9cb1dc5\"}"}]}
```

So when I finally get the logs, after pressing F5 a hundred times, I am quite frustrated.

The bottom of the job that says ERROR in the UI just reported this:

```
2025-06-18T14:35:21Z [WARNING] [/jobservice/job/impl/gc/garbage_collection.go:330]: [fc61523c-3fdf-49f0-b9ca-e6439d44c06f][198488/198489] no blob found to mark gc candidate deleting, ID:1551414, digest:sha256:2451ec8874295cf45185e73a3e05ab8705414acf97527afe93ec10077b233a83
2025-06-18T14:35:21Z [WARNING] [/jobservice/job/impl/gc/garbage_collection.go:330]: [fc61523c-3fdf-49f0-b9ca-e6439d44c06f][198489/198489] no blob found to mark gc candidate deleting, ID:1530094, digest:sha256:254f0fad0bf3d9e4e9641e8ad4872d37376e40966e7b2059236076cba2249b2e
2025-06-18T14:35:21Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:480]: 1230 blobs and 462 manifests are actually deleted
2025-06-18T14:35:21Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:481]: The GC job actual frees up 19891 MB space.
2025-06-18T14:36:05Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:518]: cache clean up completed
2025-06-18T14:36:05Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:200]: success to run gc in job.
```

And that was the job that said:

```
1230 blob(s) and 462 manifest(s) deleted, 19.43GiB space freed up
```

But claimed it could clean up way more:

```
2025-06-15T00:03:14Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:273]: 145696 blobs and 52793 manifests eligible for deletion
2025-06-15T00:03:14Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:274]: The GC could free up 2315860 MB space, the size is a rough estimation.
```

I think it might have to do with this problem: https://github.com/goharbor/harbor/issues/18014 

The job also ran for three days. GC could really use some love. 

![Image](https://github.com/user-attachments/assets/401d1cbd-704f-46f6-9f2f-016fadb6d1e4)

1. Job reports ERROR, but log does not.
2. Log has to be F5'd for ages before it shows up.
3. The estimated space to clean up is miles from the real space cleaned up.
4. Takes days to run.
5. cosigned images are not cleaned up.

Overall harbor GC is in a really bad place and needs some love. 

10 minutes: 

```
2025-06-26T18:21:26Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:339]: [96c3805e-f858-4602-8dad-89b6a3a46c64][396/257709] delete the manifest with registry v2 API: systems/asset-search/backend, application/vnd.oci.image.manifest.v1+json, sha256:42b0b04c8c80eea2484f3ada37026aca0f32b6894398025e673857d39aedfb27
2025-06-26T18:31:06Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:368]: [3114d9da-6ccf-4868-a1ec-244ef237e8c6][346/257709] delete manifest from storage: sha256:a9e56d001d48ae7aa283255a0d5184a4aff07bc2a802133b7996be05d4436c15
```

21 minutes:

```
2025-06-26T19:00:48Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:339]: [96c3805e-f858-4602-8dad-89b6a3a46c64][473/257709] delete the manifest with registry v2 API: the-data-team/dbt/dbt, application/vnd.docker.distribution.manifest.v2+json, sha256:e12e7f0adbb7780d8496ec900541088b44fad455e5a34134c8bd042b34975b48
2025-06-26T19:21:55Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:368]: [3114d9da-6ccf-4868-a1ec-244ef237e8c6][434/257709] delete manifest from storage: sha256:f109e911709268187464ffdd6c856118387a1f68afd755b44a2c14343081dc10
```

**Comment by reasonerjt** [comment]
In v2.14 timeframe we only have time to write a proposal to improve the performance of GC:
https://github.com/goharbor/community/pull/265 

Later when there's a PoC, we may ask community users in this thread to try the build.

Hence, I'm moving it out of v2.14, but keep it open in the backlog.

**Comment by wy65701436** [comment]
All the related the PR has been merged:

https://github.com/goharbor/harbor/pull/22507
https://github.com/goharbor/harbor/pull/22382

Close it by the above PRs.
